### PR TITLE
60 refactor unit code

### LINKED ISSUE #60: Refactor unit script
Change the player scripts so that they are separated into components

### PR DESCRIPTION
**Pull request by ThBrokenToaster**
**Make sure to set the pull request as a DRAFT**

Fixes #60  .

Changes proposed in this pull request:
- Refactored code and namespaced everything
- Added models and animations

### CODE REVIEW

**Review by Toeecko — DISMISSED** [review]
looks good

File: Assets/Scripts/Battlefield/CreatureScripts/AAbilitySO.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        public abstract void TriggerAbility(GameObject target);
+
+        public int Roll(int dice)
+        {
+            return Random.Range(1, dice);
+        }
+
+        public int Roll()
+        {
+            return Roll(20);
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/CreatureScripts/AAbilitySO.cs]: Are we using the dice mechanism still? i thought we abandoned the DND method with d20's and the like?

File: Assets/Scripts/Battlefield/CreatureScripts/CreatureBase.cs [hunk 2]
```diff
@@ -2,54 +2,58 @@
...
-        movement = maxMovement;
-    }
+        [Header("Creature Info")]
+        public string creatureName;
+        public int maxHealth;
+        public int health;
+        public Animator anim;
+
+        [HideInInspector]
+        public bool action = true;
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/CreatureScripts/CreatureBase.cs:16]: These flags for holding the state of the creature during its turn should probably made as a component so that we can abstract the turn information and from the creature itself. Maybe like a TurnBrain that holds whether or not an action can be taken and what can still be done. This will make it much easier to create a subclass for an AI brain vs a player brain that can be dropped in as a component and asked to deal with the "what can i do and what should i do" turn logic. 

tldr Better Abstraction between creature and its turn managing "brain".

File: Assets/Scripts/Battlefield/CreatureScripts/UniqueCreature.cs [hunk 3]
```diff
@@ -1,81 +1,75 @@
...
-    public void UseAbility(int i, GameObject target)
-    {
-        anim.SetTrigger("Attack");
-        switch (abilities[i].TypeOfActionRequired)
+        /// <summary>
+        /// This method chnages the highlighting of a unit based on an integr.  3 = no highlight, 1 = blue, 2 = red.
+        /// When highlighting a unit dont forget to set their material back to normal using glow(3). 
+        /// This function might be removed in the future in favor of using a free store asset. 
+        /// </summary>
+        public void Glow(int glow)
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/CreatureScripts/UniqueCreature.cs:40]: Probably should be taken care of by another component that determines its state and the glow it should have instead of being part of the UniqueCreature itself?

File: Assets/Scripts/Battlefield/CreatureScripts/UnitAbilitiesContainer.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            foreach (AAbilitySO ability in abilities)
+            {
+                ability.Initialize(transform.gameObject);
+            }
+        }
+
+        public void UseAbility(int i, GameObject target)
+        {
+            unit.animator.SetTrigger("Attack");
+            switch (abilities[i].TypeOfActionRequired)
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/CreatureScripts/UnitAbilitiesContainer.cs]: Instead of switch-casing here on an abilities type, it would be ideal to have subclasses for types of abilities and the ability itself can do the check for wether the action can be performed. This is more modular and allows us to react to misclicks and/or invalid ability use within the ability scriptable object itself instead of having a god-method/god-class here to deal with all situations.

File: Assets/Scripts/Battlefield/CreatureScripts/UnitBase.cs [hunk 5]
```diff
@@ -3,44 +3,49 @@
...
+            agent.destination = pos;
         }
-        agent.destination = pos;
-    }
 
-    static public int Roll(int dice)
-    {
-        return Random.Range(1, dice);
-    }
+        static public int Roll(int dice)
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/CreatureScripts/UnitBase.cs]: random algorithms should just be in some random utility class instead of being coupled to the unitbase class itself. This can just be a singleton class or a static class with tons of random related utility functions.

File: Assets/Scripts/Battlefield/PlayerInputHandler.cs [hunk 6]
```diff
@@ -1,23 +1,28 @@
...
+using SwordAndBored.Battlefield.CreaturScripts;
+using SwordAndBored.Battlefield.TurnMechanism;
+using SwordAndBored.Battlefield.CameraUtilities;
 
-
-namespace SwordAndBored.BattleMechanism
+namespace SwordAndBored.Battlefield
 {
-    public class Interact : MonoBehaviour
+    public class PlayerInputHandler : MonoBehaviour
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/PlayerInputHandler.cs]: should probably be integrated into the "TurnBrain" so that a player turnbrain subclass has this and an enemy turnbrain subclass doesn't and this will also ensure that every unit has its own instead of a master InputHandler that has to be passed around to all player units. Performance-wise I think we can take the hit without worrying too much for now since there will never be more than 6 or instances of these per battlefield.

File: Assets/Scripts/Battlefield/TurnMechanism/TurnManager.cs [hunk 7]
```diff
@@ -1,42 +1,48 @@
...
+        {
+            activePlayer = manager.NextEntity().GetComponent<UniqueCreature>();
+            text.text = "Current Player: " + activePlayer.creatureName;
+            activePlayer.StartTurn();
+        }
+
+        void Update()
         {
-            actionsLeft[0].color = new Color(1, 0, 0, .2f);
+            if (activePlayer.action)
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/TurnMechanism/TurnManager.cs]: This logic should be in the player himself and there should be something akin to activePlayer.doTurn() in here and the player holds his own state information.

**Review by subashch6 — COMMENTED** [review]
The format and refactoring is super nice compared to earlier. Most of my questions lie in the territory of future extensibility and should probably be made aware and discussed before moving ahead.

File: Assets/Scripts/Battlefield/CreatureScripts/CreatureBase.cs (hunk 2 above)
Comment (ThBrokenToaster) [reply Assets/Scripts/Battlefield/CreatureScripts/CreatureBase.cs:16]: The playerInputManager is effectively a turnBrain and if we abstracted those they would go there.  Also you've reminded me that bonus and reaction need to be deleted.   The current system keeps track of actions per player.

File: Assets/Scripts/Battlefield/CreatureScripts/AAbilitySO.cs (hunk 1 above)
Comment (ThBrokenToaster) [reply Assets/Scripts/Battlefield/CreatureScripts/AAbilitySO.cs]: Ima be honest, I just forgot to remove those methods.  My b

File: Assets/Scripts/Battlefield/CreatureScripts/UniqueCreature.cs (hunk 3 above)
Comment (ThBrokenToaster) [reply Assets/Scripts/Battlefield/CreatureScripts/UniqueCreature.cs:40]: I'm most likely going to change this so that it uses a similar system to the asset that the other team mentioned to highlight objects instead of changing the material

File: Assets/Scripts/Battlefield/CreatureScripts/UnitAbilitiesContainer.cs (hunk 4 above)
Comment (ThBrokenToaster) [reply Assets/Scripts/Battlefield/CreatureScripts/UnitAbilitiesContainer.cs]: This is no longer going to be a switch statement once the bonus and reaction bools get deleted.  It is just a leftover remnant of the old system.  Ill fix it

File: Assets/Scripts/Battlefield/TurnMechanism/TurnManager.cs (hunk 7 above)
Comment (ThBrokenToaster) [reply Assets/Scripts/Battlefield/TurnMechanism/TurnManager.cs]: The UI changes like this are going to be a part of what Rory is working on.  For now we can leave this here and Rory's UI branch will fix this and render color changes like this unnecessary

File: Assets/Scripts/Battlefield/AbstractTurnBrain.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+using SwordAndBored.Battlefield.CreaturScripts;
+using SwordAndBored.Battlefield.TurnMechanism;
+using SwordAndBored.Battlefield.CameraUtilities;
+
+
+namespace SwordAndBored.Battlefield
+{
+    public abstract class AbstractTurnBrain : MonoBehaviour
+    {
+        public TurnManager turnManager;
```
Comment (subashch6) [inline Assets/Scripts/Battlefield/AbstractTurnBrain.cs]: Since you are a component of a unit, do not have access to the turnManager but just access the player you are attached to. This will cause every brain for every unit to run UpdateBrainHandler() but on the activePlayer from the turn manager every frame. 

The modularity of having a brain for each player is lost if you ask the TurnManager for the activePlayer. See if you can architect the code so that each brain's update happens only if the player is the activePlayer. 

My suggestion would be running unit.DoTurn() in the turnManagers update() function which in fact runs myBrain.UpdateBrainHandler() along with other things like the gui and such or you can have the currently active gameobject know it is the activePlayer and tell it to take control of gui and run its brain while it is active and the turn manager can make a unit active or unactive.

**Review by subashch6 — CHANGES_REQUESTED** [review]
It seems like i wrote a lot but all i am saying is that turnManager should not have to be given to each entity or gameobject. it should be the turnManager that indicates that a player is the active player and then the activae player take control of the gui and runs its brain and the like until the client clicks end turn.

**Review by Toeecko — DISMISSED** [review]

File: Assets/Scripts/Battlefield/CreatureScripts/UnitStats.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+        public int attack;
+        public int specialAttack;
+        public int defense;
+        public int specialDefense;
+        public int speedIntit;
+        public int movement;
+
+    }
+
+}
```
Comment (wynn-rj) [inline Assets/Scripts/Battlefield/CreatureScripts/UnitStats.cs:20]: This is not different than a `GenericStats` class. Why is this included?

File: Assets/Scripts/Battlefield/Camera/CameraTarget.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+using SwordAndBored.Battlefield.TurnMechanism;
+
+namespace SwordAndBored.Battlefield.CameraUtilities
+{
+    public class CameraTarget : MonoBehaviour
+    {
+
+        public Transform player;
+
+
```
Comment (wynn-rj) [inline Assets/Scripts/Battlefield/Camera/CameraTarget.cs]: Spacing is very bad here

File: Assets/ImportsAndSampleAssets/ThirdPersonMovementAnimatior/LowMan/Scripts/ThirdPersonCharacter.cs [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+namespace UnityStandardAssets.Characters.ThirdPerson
+{
+	[RequireComponent(typeof(Rigidbody))]
+	[RequireComponent(typeof(CapsuleCollider))]
+	[RequireComponent(typeof(Animator))]
+	public class ThirdPersonCharacter : MonoBehaviour
+	{
+		[SerializeField] float m_MovingTurnSpeed = 360;
+		[SerializeField] float m_StationaryTurnSpeed = 180;
+		[SerializeField] float m_JumpPower = 6f;
```
Comment (wynn-rj) [inline Assets/ImportsAndSampleAssets/ThirdPersonMovementAnimatior/LowMan/Scripts/ThirdPersonCharacter.cs]: Is this file actually being used?

**Review by wynn-rj — COMMENTED** [review]
`Assets/TestConn.db` and its meta file should be removed as they are artifacts from the database testing

**Review by subashch6 — DISMISSED** [review]

**Review by Toeecko — DISMISSED** [review]

**Review by Toeecko — APPROVED** [review]

**Review by subashch6 — APPROVED** [review]
Yeet
